ramips: add support for D-Link DAP-1325-A1

spleen1981 opened the pull request:
This patch adds support for D-Link DAP-1325-A1 (Range Extender Wi-Fi N300)
	
### Specifications:

- SoC: 580Mhz MT7628NN
- RAM: 64MB, DDR2 SDRAM
- Storage: 8MB, SPI
- Ethernet: 1x 10/100 LAN port
- WIFI: 2.4 GHz 802.11bgn
- LED: Status (2x to provide 3 colors), Wi-Fi Signal Strength (4x)
- Buttons: Reset, WPS
- UART: Serial console (57600, 8n1)
Row of 4 holes near LAN port, starting from square hole:
3.3V, TX,RX,GND
- FCC ID: fccid.io/KA2AP1325A1/

### Installation:

### Failsafe UI
Firmware can be uploaded with Failsafe UI web page:
- turn device off
- press and hold reset button
- turn device on
- keep holding reset until red wifi strength led turns on (ab. 10sec)
- connect to device through LAN port
PC must be configured with static ip (192.168.0.x)
- connect to 192.168.0.50
- select image to be flashed and upload.
Device will reboot after successful update

### Serial port/TFTP server

- Connect through serial connectors on PCB (e.g. with teraterm)
- Set up a TFTP server, and connect through LAN with static IP
- Put image file in the root of the server
- Boot the device and select '2' at U-Boot startup
- Set device IP, server IP and image file name
- Start upload and flash

Signed-off-by: Giovanni Cascione ing.cascione@gmail.com

s-2 commented:
Great to see efforts to support these devices!

Please re-check https://openwrt.org/submitting-patches for the contents of the commit message, and squash changes into a single commit.

Is any further magic required to set wifi mac address correctly (to match the one printed on the label, or whatever the OEM firmware uses), or will it just work with the defaults?

Factory image crypto seems somewhat straightforward on this one, compared to other D-Link devices (it's apparently AES in ECB mode, key `3bae351628aed2a60bf71528c9cfdf3c`, no iv, for everything after the `0x40` bytes custom header), but still I'm not sure whether this would be ultimately accepted by maintainers; I guess flashing via recovery will be fine for now :slightly_smiling_face:

spleen1981 commented:
@s-2  commit message is updated to try to comply as much as possible with the guidelines, please let me know if you still see something specific to be fixed.
Commits also squashed into one.

Single MAC shown on original printed label is stored as it is in 'factory + 0x28'.

musashino205 commented:
commit message is still missing, PR comment != commit message.
And you need to add a prefix to the commit title (```ramips: ```).

s-2 commented:
I'm not very familiar with the ramips target, but from looking at other single-port devices, the board might still need to be added to `ramips_setup_interfaces()` in `02_network` (probably in the first case block, disabling the switch and setting `eth0` to `LAN`).

Since this device has LEDs for signal strength indication, you might be interested in the `rssileds` package.
Check e.g. DAP-1330 from the `ath79` target, RSSI thresholds are declared in `01_leds`:
https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/generic/base-files/etc/board.d/01_leds

spleen1981 committed:
ramips: add support for D-Link DAP-1325-A1

This patch adds support for D-Link DAP-1325-A1 (Range Extender Wi-Fi N300)

Specifications:
- SoC: 580Mhz MT7628NN
- RAM: 64MB, DDR2 SDRAM
- Storage: 8MB, SPI (W25Q64JVSSIQ)
- Ethernet: 1x 10/100 LAN port
- WIFI: 2.4 GHz 802.11bgn
- LED: Status (2x to provide 3 colors), Wi-Fi Signal Strength (4x)
- Buttons: Reset, WPS
- UART: Serial console (57600, 8n1)
Row of 4 holes near LAN port, starting from square hole:
3.3V, TX,RX,GND
- FCC ID: fccid.io/KA2AP1325A1/

Installation:

Failsafe UI
Firmware can be uploaded with Failsafe UI web page:
- turn device off
- press and hold reset button
- turn device on
- keep holding reset until red wifi strength led turns on (ab. 10sec)
- connect to device through LAN port
PC must be configured with static ip (192.168.0.x)
- connect to 192.168.0.50
- select image to be flashed and upload.
Device will reboot after successful update

Serial port/TFTP server
- Connect through serial connectors on PCB (e.g. with teraterm)
- Set up a TFTP server, and connect through LAN with static IP
- Put image file in the root of the server
- Boot the device and select '2' at U-Boot startup
- Set device IP, server IP and image file name
- Start upload and flash

Signed-off-by: Giovanni Cascione <ing.cascione@gmail.com>

spleen1981 commented:
@musashino205 thanks, should be ok now.

@s-2 with no modifications to `ramips_setup_interfaces()` in `02_network` it seems the device is working as intended (switch disabled and eth0 working properly). Not sure if there are other reasons to expressly add the device in one case block or not.
Regarding `rssileds` my feeling is that it would be better to leave to the user the choice of defining leds behaviour as required, but I may be wrong,

adschm commented on target/linux/ramips/dts/mt7628an_dlink_dap-1325-a1.dts:
```diff
@@ -0,0 +1,125 @@
+		
+		
+		wifi-verylow {
+			label = "red:wifi-low";
```
red:wifi-verylow for consistency ...

spleen1981 replied:
fixed

adschm commented on target/linux/ramips/dts/mt7628an_dlink_dap-1325-a1.dts:
```diff
@@ -0,0 +1,125 @@
+
+&state_default {
+	gpio {
+		groups = "wdt";
```
There are several groups missing here ...

spleen1981 replied:
updated

adschm commented on target/linux/ramips/dts/mt7628an_dlink_dap-1325-a1.dts:
```diff
@@ -0,0 +1,125 @@
+
+&wmac {
+	status = "okay";
+	mtd-mac-address = <&factory 0x28>;
```
Address in 0x4 is wrong/dummy?

spleen1981 replied:
no valid address in 0x4 (00 00 22 00 00 00)

adschm commented on target/linux/ramips/dts/mt7628an_dlink_dap-1325-a1.dts:
```diff
@@ -0,0 +1,125 @@
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
```
This can go faster ...

spleen1981 replied:
Right, it's W25Q64JVSSIQ.
Increased to 50MHz as per datasheet (Clock frequency for Read Data instruction - MAX)

adschm commented on target/linux/ramips/dts/mt7628an_dlink_dap-1325-a1.dts:
```diff
@@ -0,0 +1,125 @@
+		reg = <0>;
+		spi-max-frequency = <10000000>;
+
+		partitions: partitions {
```
DT label is not required here.

spleen1981 replied:
fixed

adschm commented on target/linux/ramips/dts/mt7628an_dlink_dap-1325-a1.dts:
```diff
@@ -0,0 +1,125 @@
+				read-only;
+			};
+
+			firmware_part: partition@50000 {
```
DT label is not required here.

spleen1981 replied:
fixed

adschm commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -100,6 +100,14 @@ define Device/d-team_pbr-d1
+  IMAGE_SIZE := 7872k
+  DEVICE_VENDOR := D-Link
+  DEVICE_MODEL := DAP-1325 A1
+  SUPPORTED_DEVICES += dap-1325-a1
```
this line can be dropped, unless you need for factory flashing ...

spleen1981 replied:
fixed

adschm commented:
This also lacks ethernet setup and maybe rssiled setup ...

spleen1981 commented:
> 
> 
> This also lacks ethernet setup and maybe rssiled setup ...

both updated

adschm commented:
Thanks, I will merge this later tonight.
